HOISTING.md

```diff
@@ -2,36 +2,50 @@
 
 Unlike decorating function expressions, decorating function declarations exposes us to a TDZ issue due to the hoisting that function declarations are subject to.
 
-### Suggestions
+## Possible Solutions
 
-#### Mirrors
+### Mirrors
 
 The TDZ issue can be resolved using [Mirrors](https://gist.github.com/rbuckton/8e6806fb6852b50e4052/) and if this proves to be an acceptable solution this proposal can be expanded to include function declarations as well.
 
-#### Decorated function execution would not be hoist until first call. ([by @hax](https://github.com/tc39/proposal-decorators/issues/40#issuecomment-369867557))
+### New syntax without hoisting
 
-```javascript
+```js
 @deco
-function f() {
-  funcBody
-}
+let f() {}
+
+@deco
+const g() {}
 ```
 
-Could be transpiled to:
+### Lazy evaluation ([by @hax](https://github.com/tc39/proposal-decorators/issues/40#issuecomment-369867557) and [revised by @rbuckton](https://github.com/tc39/proposal-decorators/issues/40#issuecomment-370010647))
+
+Make decorator evaluation lazy and apply it at one of two times (whichever comes first):
+
+- When the function declaration is encountered during evaluation,
+- The first time the function declaration is accessed as a value.
 
 ```javascript
-function f(...args) {
-  $do_decorate_f()
-  return $decorated_f.call(this, ...args)
+let g = f; // access f before declaration
+f(); // invoke f before declaration
+
+@deco function f() {
+  f(); // invoke f inside of declaration
 }
+```
 
-var $decorated_f
+Could be transpiled to:
 
-function $do_decorate_f() {
-  if (!$decorated_f) $decorated_f = deco(function f() {
-    funcBody
-  })
-}
+```javascript
+let g = $$get_f(); // access f before declaration
+$$get_f()(); // invoke f before declaration
 
-$do_decorate_f()
+function f() {
+  $$get_f()(); // invoke f inside of declaration
```

Comment:
This line differs from the original code `f()` of @rbuckton which will call the undecorated `f`. Not sure which semantic is more correct.